bundles/paikkatietoikkuna/coordinatetransformation/components/CoordinateSystemSelection.js

```diff
@@ -0,0 +1,241 @@
+            this.element = el;
+        },
+        getElement: function () {
+            return this.element;
+        },
+        createUi: function () {
+            var wrapper = this._template.systemWrapper.clone();
+
+            var coordinateSystemSelection = this._template.coordinateSystemSelection({
+                title: this.loc.coordinatesystem.title,
+                geodetic_datum: this.loc.coordinatesystem.geodetic_datum,
+                coordinate_system: this.loc.coordinatesystem.coordinate_system,
+                map_projection: this.loc.coordinatesystem.map_projection,
+                geodetic_coordinate_system:this.loc.coordinatesystem.geodetic_coordinatesystem,
+                elevation_system:this.loc.coordinatesystem.heigth_system 
+            });
+            wrapper.append(coordinateSystemSelection);
+            this.setElement(wrapper);
+            this.populateSelect();
+            this.handleInfoLink();
+        },
+        /**
+         * @method populateSelect
+         * @desc creates an instance of the { Oskari.coordinatetransformation.component.select },
+         * and fills it with data
+         */
+        populateSelect: function () {
+            var me = this;
+            var wrapper = this.getElement();
+            this.select.create();
+            this.selectInstance = this.select.getSelectInstances();
+            this.dropdowns = this.select.getDropdowns();
+            Object.keys( this.dropdowns ).forEach( function( key ) {
+                jQuery( wrapper.find( '.transformation-system' ).find( me.makeClassSelector(key) ).append( me.dropdowns[key] ));
+            });
+            this.handleSelectionChanged();
+        },
+        handleInfoLink: function () {
+            var me = this;
+            this.getElement().find('.infolink').on('click', function ( event ) {
+                event.stopPropagation();
+                var key = this.parentElement.parentElement.dataset.system;
+                me.systemInfo.show( jQuery( this ), key );
+            });
+        },
+        handleSelectionChanged: function () {
+            var systems =  this.getElement().find('.system');
+            var me = this;
+            systems.each(function (index) {
+                jQuery( this ).on('change', function ( e ) {
+                    me.handleSystemChange(e);
+                });
+            })
+        },
+        handleSystemChange: function (e) {
+            var current = this.selectInstance[e.currentTarget.dataset.system];
+            var currentValue = current.getValue();
+            this.handleSelectValueChange(currentValue);
+        },
+        /**
+         * @method handleDropdownOptions
+         * @param {string} valueClass - class selector to show options for 
+         * @param {string} keyToEmpty - optional param to empty only one specific key in the dropdown 
+         */
+        handleDropdownOptions: function (valueClass, keyToEmpty) {
```

Comment:
If I understand this correctly the dropdowns have multiple sets of options. Sometimes all the options are shown but sometimes some of them are hidden from the user. Option-tags have a class that is used to identify sets of options. Options belonging to a given set (valueClass) can be easily shown after all options have been hidden. "keyToEmpty" is used if only options from one dropdown needs to be hidden before showing all options in all the dropdowns having valueClass. So it assumes the other dropdowns have the non-valueClass options hidden already OR it doesn't matter if they are hidden as long as the valueClass options are visible?

Could this be written more clearly as updateDropdownOptions(dropdownId, valueSetId) where it only ever updates a single dropdown, dropdown is key/keyToEmpty from this method and valueSetId is valueClass? When you need to do a full update you could do:

                Object.keys( dropdowns ).forEach( function ( key ) {
                    // for showing all options don't send valueSetId parameter
                    me.updateDropdownOptions(key, valueSetId); 
                });